src/main/java/org/damap/base/rest/invenio_damap/InvenioDAMAPService.java

```diff
@@ -0,0 +1,110 @@
+    this.dmpService = dmpService;
+    this.versionService = versionService;
+  }
+
+  // Resolves user identity based on the DMPs and returns dmp list and list of
+  // identifiers if successful.
+  public SimpleEntry<List<DmpDO>, List<String>> resolveDmpsAndIds(JsonObject identifiers) {
+    List<DmpDO> personDmpList = null;
+    List<String> matchingIdentifiers = new ArrayList<>();
+
+    for (String key : identifiers.keySet()) {
+      String identifier = identifiers.getString(key);
+      List<DmpDO> dmps = dmpService.getDmpDOListByPersonId(identifier);
+
+      // Check if returned DMPs is null or the list is empty.
+      // If yes, no resolving can take place, move to the next identifier.
+      if (dmps == null || dmps.isEmpty()) continue;
+
+      // Here dmps would contain at least one DMP so keep track of the current
+      // iterated identifier.
+      matchingIdentifiers.add(identifier);
+
+      // Keep track of the 1st list that matches at least one DMP.
+      // Then compare it with other non-null or non-empty lists.
+      if (personDmpList == null) personDmpList = dmps;
+
+      // If a DMP list is different than the previous one, a different identity was
+      // resolved.
+      else if (!(personDmpList.containsAll(dmps) && dmps.containsAll(personDmpList))) {
+        throw new UnauthorizedException("Mismatch in resolved identities.");
+      }
+    }
+
+    // Check if DMP list is null. This means that no resolving was possible.
+    if (personDmpList == null) throw new NotFoundException("No valid identities were found.");
+
+    return new SimpleEntry<>(personDmpList, matchingIdentifiers);
+  }
+
+  @Transactional
+  public DmpDO addDataSetToDMP(long dmpId, Dataset dataset) {
+
+    DmpDO dmpDO = dmpService.getDmpById(dmpId);
+    var datasetDO =
+        dmpDO.getDatasets().stream()
+            .filter(
+                ds -> {
+                  var localIdentifier = ds.getDatasetId();
+                  var externalIdentifier = dataset.getDatasetId();
+
+                  if (localIdentifier == null || externalIdentifier == null) {
+                    return false;
+                  }
+
+                  return localIdentifier.getIdentifier() != null
+                      && externalIdentifier.getIdentifier() != null
+                      && localIdentifier.getType() != null
+                      && externalIdentifier.getType() != null
+                      && localIdentifier.getIdentifier().equals(externalIdentifier.getIdentifier())
+                      && localIdentifier
+                          .getType()
+                          .toString()
+                          .equalsIgnoreCase(externalIdentifier.getType().name());
+                })
+            .findFirst()
+            .orElse(new DatasetDO());
+
+    datasetDO = InvenioDamapResourceMapper.mapMaDMPDatasetToDatasetDO(dataset, datasetDO, dmpDO);
+
+    dmpDO.getDatasets().add(datasetDO);
+    dmpDO = dmpService.update(dmpDO);
```

Comment:
If `.findFirst()` returns a datasetDO, it is readded to the dmpDO datsasets. This leads to a duplicated dataset in the list.

Reply:
Yes, i was so preoccupied with the mapper i forgot to address my own comment 😅 
Thanks for reminding me.